### ISSUE TITLE
[test_pow_fractional.py:49] | (Runtime) Integration Tests | runtime-integration-tests / runtime_fd_python_4 [bh_p150b_civ2]

### ISSUE THREAD

**Issue by ebanerjeeTT** [tenstorrent/tt-metal#48053]
## CI Failure Report

**Summary:** [CI] test_pow_fractional_composite PCC accuracy check fails in runtime_fd_python_4 [bh_p150b_civ2]

**Workflow:** (Runtime) Integration Tests
**Job:** runtime-integration-tests / runtime_fd_python_4 [bh_p150b_civ2]

### Regression timeline

- **Last passing commit:** N/A
- **First failing commit:** `ae414c18e46f` — 2026-06-24T06:20:46Z — [job](https://github.com/tenstorrent/tt-metal/actions/runs/28079421656/job/83140314817)
- **Most recent failing commit:** `1a118dcad608` — 2026-06-25T06:23:30Z — [job](https://github.com/tenstorrent/tt-metal/actions/runs/28151153934/job/83377688864)

### Failing job URLs (last 2 runs)

- https://github.com/tenstorrent/tt-metal/actions/runs/28151153934/job/83377688864
- https://github.com/tenstorrent/tt-metal/actions/runs/28079421656/job/83140314817

### Failing test path(s)

- `tests/tt_eager/python_api_testing/unit_testing/misc/test_pow_fractional.py::test_pow_fractional_composite`

Run command:
```
pytest tests/tt_eager/python_api_testing/unit_testing/ -xvvv --splits 7 --group 4
```

### Error excerpt (terminal failure from logs)

**Run 1 (commit `1a118dcad608`, 2026-06-25):**
```
________________________ test_pow_fractional_composite _________________________

2026-06-25 07:30:31.372 | INFO | tests.tt_eager.python_api_testing.unit_testing.misc.test_pow_fractional:test_pow_fractional_composite:48 - 0.6391644024467982
>       assert passing
E       assert False

tests/tt_eager/python_api_testing/unit_testing/misc/test_pow_fractional.py:49: AssertionError

FAILED tests/tt_eager/python_api_testing/unit_testing/misc/test_pow_fractional.py::test_pow_fractional_composite - assert False
========== 1 failed, 26 passed, 5533 deselected, 3 warnings in 21.69s ==========
```

**Run 2 (commit `ae414c18e46f`, 2026-06-24):**
```
________________________ test_pow_fractional_composite _________________________

2026-06-24 07:38:31.514 | INFO | tests.tt_eager.python_api_testing.unit_testing.misc.test_pow_fractional:test_pow_fractional_composite:48 - 0.6391644024467982
>       assert passing
E       assert False

tests/tt_eager/python_api_testing/unit_testing/misc/test_pow_fractional.py:49: AssertionError

FAILED tests/tt_eager/python_api_testing/unit_testing/misc/test_pow_fractional.py::test_pow_fractional_composite - assert False
========== 1 failed, 26 passed, 5533 deselected, 3 warnings in 20.66s ==========
```

### Reproduction steps

```bash
pytest tests/tt_eager/python_api_testing/unit_testing/misc/test_pow_fractional.py::test_pow_fractional_composite -xvvv
```

Or to reproduce the exact split context:
```bash
pytest tests/tt_eager/python_api_testing/unit_testing/ -xvvv --splits 7 --group 4
```

Hardware target: `bh_p150b_civ2` (Blackhole)

### Notes

The test `test_pow_fractional_composite` computes a fractional power using `ttnn` ops (log, exp, mul, pow) and compares the result against a PyTorch reference using PCC (Pearson Correlation Coefficient). The PCC value of `0.6391644024467982` is well below the passing threshold, indicating a numerical accuracy regression in the composite fractional power computation.

Notably, both runs produce **the exact same PCC value** (`0.6391644024467982`), which is consistent with a deterministic numerical error introduced by a code change rather than a flaky/intermittent failure. The test uses a fixed random seed (`torch.manual_seed(577215)`).

**Determinism check:**
- Run 1 (`1a118dcad608`): `FAILED tests/tt_eager/python_api_testing/unit_testing/misc/test_pow_fractional.py::test_pow_fractional_composite - assert False` — PCC = `0.6391644024467982`
- Run 2 (`ae414c18e46f`): `FAILED tests/tt_eager/python_api_testing/unit_testing/misc/test_pow_fractional.py::test_pow_fractional_composite - assert False` — PCC = `0.6391644024467982`

`Auto-triage-workflow: (Runtime) Integration Tests`
`Auto-triage-job-name: runtime-integration-tests / runtime_fd_python_4 [bh_p150b_civ2]`

## About this issue (automated CI auto-triage)
This issue was opened and assigned by an automated AI system. It posts through a bot token, so notifications may look like they come from a person, but they are automated.

The owner is selected from recent commit history and job-ownership data. It reflects who is *likely able to fix or route* this, not necessarily who caused it, so it can be imperfect.

**If this was assigned to the wrong person, you can:**
- Comment `REASSIGN OWNER: <why this is outside your area / which team or test-area it belongs to>` and the system will automatically re-route it to a better-suited owner.
- Unassign yourself (this stops notifications to you).
- Close the issue if it is not a real or relevant failure.

The assigned owner is reminded about every 3 days until the issue is resolved. Full guide: https://tenstorrent.atlassian.net/wiki/spaces/MI6/pages/2545221692/Auto-Triage+Info

## Pipeline report
Analysis report: not found
Slack notification: not found

**Comment by Aswinmcw** [comment]
## Analysis: `test_pow_fractional_composite` PCC failure

*Reproduces on both WH N150 (civ2) and BH P150b (civ2) — PCC ≈ 0.639 vs threshold 0.99*

---

### Test overview

`test_pow_fractional_composite` computes:
```python
pow_frac_composite = exp(log(x, fast_and_approximate_mode=True) * y_frac) * pow(x, y_floor)
```

The input `x = torch.randn((1, 2, 32, 32))` contains ~50% negative values. For negative `x`, `log(x)` → NaN, propagating NaN through the whole expression. The golden `torch.pow(x, y)` also returns NaN for negative non-integer powers. A PCC of 0.639 (vs 0.99) indicates **systematic error across the positive-value elements**, not just a NaN-propagation mismatch.

---

### Auto-triage first-failing-commit is a red herring

The auto-triage reports `ae414c18e46f` (June 23, rotary embedding RISC-V cleanup) as the first failing commit — but it only touches rotary embedding kernels and has no interaction with eltwise/log/exp ops. "Last passing commit: N/A" means no passing run was found in the monitoring window; the regression started before the triage window. Look earlier.

---

### Root cause candidates

#### Candidate 1 (PRIMARY) — `cec422302cbd` · "RoundMode clarification (tt_metal/hw) 3/6" · June 15

PR #46851 changed `ckernel_sfpu_log.h` for **both WH and BH**:

```cpp
// BEFORE:
e_float = sfpi::convert<sfpi::vFloat>(abs_e, sfpi::RoundMode::NearestEven);
result  = sfpi::convert<sfpi::vFloat16b>(result, sfpi::RoundMode::NearestEven);

// AFTER (current):
e_float = sfpi::convert<sfpi::vFloat>(abs_e, sfpi::RoundMode::Nearest);
result  = sfpi::convert<sfpi::vFloat16b>(result, sfpi::RoundMode::Nearest);
```

The PR is described as "backwards compatible" because WH/BH hardware rounds nearest-away-from-zero. However, SFPI was bumped 7.58.0 → 7.59.0 → 7.60.0 between June 15–17. If SFPI 7.59.0/7.60.0 emits a different hardware instruction for `RoundMode::Nearest` than for `RoundMode::NearestEven`, the actual rounding behaviour of the BF16 conversion (`result = convert<vFloat16b>(result, ...)`) changes. A 1-ULP difference in log results would not cause PCC 0.639 alone, but incorrect exponent selection in the polynomial path would.

The fast-approximate path (used by this test) also skips normalisation:

```cpp
if constexpr (!FAST_APPROX) {
    a = a * sfpi::vConst1 + sfpi::vConst0;  // flush -0.0 / subnormals → +0.0
}
```

Without normalisation, edge-case inputs interact differently with the changed RoundMode conversion.

#### Candidate 2 (SECONDARY) — `987c7451ded8` · "Remove dependency on debug bit 11 for typecast" · June 21

PR #46231 removed debug bit 11 management from `_llk_math_eltwise_sfpu_uninit_()` on both WH and BH. Previously, after every SFPU op with FP32 dest accumulation enabled, uninit re-engaged the HW bug workaround:

```cpp
// REMOVED:
if (cfg_read(ALU_ACC_CTRL_Fp32_enabled_ADDR32) & ALU_ACC_CTRL_Fp32_enabled_MASK)
{
    _llk_math_dbg_feature_enable_();  // re-enable bit-11 workaround for next op
}
```

Debug bit 11 (`RISCV_DEBUG_REG_DBG_FEATURE_DISABLE`) is the FPU HW bug workaround from issue #46219. After this PR, if any preceding operation in the same test session leaves bit 11 in the wrong state, subsequent log/exp SFPU ops run without the workaround — potentially exercising the underlying FPU bug. This commit landed June 21, two days before the first failing CI run.

---

### Suggested investigation steps

1. **Bisect `cec422302cbd`** — Revert `RoundMode::Nearest` → `RoundMode::NearestEven` in `ckernel_sfpu_log.h` (WH + BH) and re-run the test. If PCC recovers, check whether SFPI 7.60.0 emits different hardware instructions for the two enum values.

2. **Bisect `987c7451ded8`** — Restore the bit-11 re-enable in `_llk_math_eltwise_sfpu_uninit_()` and re-run. If PCC recovers, determine whether log/exp require the bit-11 workaround and whether they should have been included in the PR's scope alongside typecast and reduce.

3. **Isolate the positive-value path** — Change the test input to avoid negatives (`torch.clamp(x, min=1e-3)`) and re-run. If PCC is still low, the error is entirely in the log polynomial computation (pointing to `cec422302cbd`). If PCC recovers to near-1.0, the error was in NaN handling or propagation.

4. **Verify fast vs full approx** — Re-run with `fast_and_approximate_mode=False` to confirm whether the regression is fast-path-specific.

---

### Fix recommendation

- **If `cec422302cbd` is the cause**: revert to `RoundMode::NearestEven` in `ckernel_sfpu_log.h` for both architectures, or confirm with the SFPI team that `Nearest` and `NearestEven` produce identical instructions in SFPI 7.59.0+.
- **If `987c7451ded8` is the cause**: explicitly manage bit 11 state in the log/exp SFPU init path (similar to the old typecast handling), or file a follow-up to verify that no SFPU op other than typecast/reduce required bit 11 management.
- **Test hardening**: the test applies fast-approximate log to a tensor with ~50% negative values from an unfixed random seed. Consider adding `torch.clamp(x, min=0.001)` or using `torch.abs(x)` as a pre-condition to restrict testing to the valid domain, and separately test NaN propagation explicitly.

cc @JasonDTX (PR #46851 author) @amahmud (PR #46231 author)

**Comment by mcw-anasuya** [comment]
**Reason for Failure:** composite decomposition path. In [line 35](https://github.com/tenstorrent/tt-metal/blob/699655e7aee849671d1242df950fb93757c94c71/tests/tt_eager/python_api_testing/unit_testing/misc/test_pow_fractional.py#L35), `yt_trunc` is a [1,1,1,1] torch tensor `tensor([[[[0.5371]]]])`, not a Python scalar. So `ttnn.multiply(log(x), yt_trunc)` has no overload that accepts a torch.Tensor for input_tensor_b (there is no implicit torch.Tensor → ttnn.Tensor conversion), so the call binds to the tensor–scalar overload, whose signature is:
`multiply(input_tensor_a: ttnn.Tensor, input_tensor_b: int | int | float, ...)`
That int | int | float is `ScalarVariant = std::variant<uint32_t, int32_t, float>`, introduced for multiply/divide in https://github.com/tenstorrent/tt-metal/pull/45984, which changed the param from a plain `float`.
As a result, nanobind resolves a `std::variant` by trying its alternatives left-to-right and the integer alternatives (uint32_t, int32_t) come before float. A 1-element float tensor satisfies the integer conversion via truncation (int(torch.tensor(0.537)) == 0), so the scalar is silently taken as 0 instead of 0.537. Earlier, the parameter was float, so the same 1-element tensor converted through the float protocol to 0.537 and the test passed. This is why a deterministic regression first appears in this [integration test](https://github.com/tenstorrent/tt-metal/actions/runs/27666415743/job/81822402131#step:7:11787) 

The following change fixes the issue by extracting the scalar: 
```
- yt_trunc = y.item() - yt_floor
+ yt_trunc = (y.item() - yt_floor).item()
```

**Comment by mcw-anasuya** [comment]
Fix merged to main https://github.com/tenstorrent/tt-metal/pull/48227
Closing this issue.
